### ISSUE TITLE
Line layer uploaded to GeoNode is rendered as point layer

### ISSUE THREAD

**Issue by capooti** [GeoNode/geonode#4827]
## Expected Behavior

A line shapefile is uploaded to GeoNode. The layer is correctly registered in GeoNode, a line style is created in GeoServer and the style is associated to the layer as the default style.

## Actual Behavior

The line layer is uploaded in GeoNode. In GeoServer two styles are created for the layer, one named "layername" (with a point symbolizer) and the other "layername_layer" (with a line symbolizer). The incorrect one is associated to the layer.

## Steps to Reproduce the Problem

I can't reproduce this on GeoNode master, but I noticed this happening randomly on two different GeoNode instances.

## Specifications

  - GeoNode version: master
  - Installation method (manual, GeoNode Docker, SPCGeoNode Docker): manual
  - Platform: Ubuntu
  - Additional details:

**Comment by capooti** [comment]
I believe this is connected with this issue as well: https://github.com/GeoNode/geonode/issues/2932

Under some circumstances GeoNode can't get the right response from GeoServer REST styles endpoint.

**Comment by afabiani** [comment]
@capooti yes, we should somehow allow GeoNode to double check if the style has been created or not.
Probably a simple upgrade of the gsconfig library would be sufficient.

FWI, we forked from the Boundless one. We now rely on a new one which runs with Python3 also:

https://github.com/GeoNode/geoserver-restconfig

Feel free to track issues and open PRs there.

**Comment by capooti** [comment]
Yes, I am aware of the fork and was already looking the source code. Should I close the ticket here and open a new one there?

**Comment by capooti** [comment]
Also, I am using already the very latest version (gsconfig==1.0.10) as I created the virtualenv using requirements.txt from master

**Comment by afabiani** [comment]
No, let's keep this issue open too. Let me know if you need some help.

**Comment by capooti** [comment]
@afabiani I can work with this but I am not sure what would be the best approach. 
How about using requests.packages.urllib3.util.retry.Retry as suggested here? https://www.peterbe.com/plog/best-practice-with-retries-with-requests

This could be done in geoserver-restconfig in catalog.create_style. Right now create_style doesn't return anything. We could return the style itself or None in case even with Retry the style can't be retrieved successfully

**Comment by afabiani** [comment]
Yep, it looks to me a good approach.

**Comment by capooti** [comment]
@afabiani I realized my virtualenv had both gsconfig==1.0.10 and geoserver-restconfig==1.0.3 and was using gsconfig. Just after uninstalling gsconfig GeoNode started using geoserver-restconfig and this issue seems gone :)
For now I am closing it, hopefully we won't need to re-open it.
Thanks again for helping!

**Comment by capooti** [comment]
A user reported that now layers are correctly styled, but thumbnails still are created using a point style. Not sure if this is connected.

**Comment by capooti** [comment]
Unfortunately I have to reopen this. The problem arise intermittently both with line and polygon layers, which are styled with lines...

**Comment by capooti** [comment]
@afabiani I debugged the issue with a script using geoserver-config and it definitely look to be a GeoServer scalability problem.

Basically in the instance where I am experiencing the problem I can successfully read a newly created style only after 4/5 seconds (!!!) from the creation of the style itself. Looking at the GeoServer log it appears that, during the style read operation, GeoFence is checking every single layer in the instance - that is why it takes a while. 

For each layer this is what I see in the log:

```
2019-09-20 02:10:21,501 DEBUG [org.geoserver.geofence] - Admin auth for User: Workspace:geonode: false
2019-09-20 02:10:21,501 DEBUG [org.geoserver.geofence] - Getting access limits for Layer a__5a0
2019-09-20 02:10:21,501 DEBUG [org.geoserver.geofence] - Getting access limits for Resource a__5a0
2019-09-20 02:10:21,501 DEBUG [org.geoserver.geofence] - ResourceInfo filter: RuleFilter[user:DEFAULT role:ANY inst:name+:default-gs ip:"127.0.0.1"+ serv:"WMS"+ req:"GETCAPABILITIES"+ ws:"geonode"+ layer:"a__5a0"+]
2019-09-20 02:10:21,501 DEBUG [org.geoserver.geofence.cache] - Request for RuleFilter[user:DEFAULT role:ANY inst:name+:default-gs ip:"127.0.0.1"+ serv:"WMS"+ req:"GETCAPABILITIES"+ ws:"geonode"+ layer:"a__5a0"+]
2019-09-20 02:10:21,501 DEBUG [org.geoserver.geofence] - Returning mode HIDE for resource FeatureTypeInfoImpl[a__5a0]
2019-09-20 02:10:21,501 DEBUG [org.geoserver.geofence] - Returning VectorAccessLimits [readAttributes=null, writeAttributes=null, writeFilter=Filter.EXCLUDE, readFilter=Filter.EXCLUDE, mode=HIDE] for layer geonode:a__5a0 and user null
2019-09-20 02:10:21,501 DEBUG [org.geoserver.geofence] - Getting access limits for workspace geonode
2019-09-20 02:10:21,501 DEBUG [org.geoserver.geofence] - Getting admin auth for Workspace geonode
2019-09-20 02:10:21,501 DEBUG [org.geoserver.geofence] - AdminAuth filter: RuleFilter[user:DEFAULT role:ANY inst:name+:default-gs ip:"127.0.0.1"+ serv:ANY req:ANY ws:"geonode"+ layer:ANY]
```

The big latency between the the style creation and its successful read (4/5 seconds in this case) explains why now GeoNode defaults to a point style. 

Any suggestion about the approach for tackling this problem?

**Comment by simboss** [comment]
Ciao @capooti,
can you share this script so that we can reproduce?

**Comment by capooti** [comment]
sure @simboss, it is really basic :)

```python
import time
from geoserver.catalog import Catalog

name = 'test999'

sld = """
<StyledLayerDescriptor version="1.0.0" xmlns="http://www.opengis.net/sld" xmlns:ogc="http://www.opengis.net/ogc"
  xmlns:xlink="http://www.w3.org/1999/xlink" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
  xsi:schemaLocation="http://www.opengis.net/sld http://schemas.opengis.net/sld/1.0.0/StyledLayerDescriptor.xsd">
  <NamedLayer>
    <Name>test999</Name>
    <UserStyle>
    <Name>test999</Name>
    <Title>test999</Title>
      <FeatureTypeStyle>
        <Rule>

<PointSymbolizer>
  <Graphic>
    <Mark>
      <WellKnownName>circle</WellKnownName>
      <Fill>
        <CssParameter name="fill">#008800</CssParameter>
      </Fill>
      <Stroke>
        <CssParameter name="stroke">#bbffbb</CssParameter>
      </Stroke>
    </Mark>
    <Size>10</Size>
  </Graphic>
</PointSymbolizer>

        </Rule>
      </FeatureTypeStyle>
    </UserStyle>
  </NamedLayer>
</StyledLayerDescriptor>
"""


url = 'http://hostname/geoserver/rest'
user = 'admin'
password = '***'
cat = Catalog(url, user, password)

cat.create_style(name, sld, overwrite=True,
                raw=True, workspace='geonode')


style = None
i = 0
while not style:
    style = cat.get_style(name, workspace='geonode')
    i += 1
    print('new request after %i seconds spent' % i)
    time.sleep(i)

print('Finally got a style!')
```

in my case it takes always 4/5 seconds before I can successfully get the new style from a request.
I was thinking: maybe there is some cache which must be invalidated - which would explain the issue?

**Comment by simboss** [comment]
Which GS version are you using?
How many layers do you have?

Is this instance accessible to us, just in case?

**Comment by capooti** [comment]
It is using latest GeoNode master, therefore GS 2.15.2
There are almost 700 layers there
If needed to debug better, I can provide you with ssh access to the instance.

**Comment by capooti** [comment]
hey @simboss did you have time to check this? looking forward to read your ideas about it

**Comment by capooti** [comment]
@afabiani unfortunately  as soon as I started using this in our instances the issues started happening again.
In the last 7/8 days we used this [1], and we didn't notice anymore #4827 and #5049 happening

[1] https://github.com/capooti/geonode/tree/camp-before-2.10.1-merge

**Comment by afabiani** [comment]
@capooti I'll compare the branches and do some tests tomorrow.

**Comment by capooti** [comment]
thanks @afabiani ;)

**Comment by capooti** [comment]
Unfortunately this is happening again, so I have to re-open it.
Everything works correctly with this branch:
https://github.com/capooti/geonode/tree/camp-before-2.10.1-merge

**Comment by capooti** [comment]
@afabiani here is a fix
https://github.com/capooti/geonode/pull/new/fixes-4827-geonode-2.10.1

the two lines which I removed are not anymore needed
cat.create_style will raise an error as the style is already there and in any case it is useless as we don't read the style again after creating it, so the next "if style" block won't be execute.

Not sure where to send the PR now that we released 2.10.1...

**Comment by afabiani** [comment]
@capooti this patch will go into 2.10.2

**Comment by afabiani** [comment]
@capooti I'm going to create a new issue referring to this one and assigned to 2.10.2 milestone.

**Comment by afabiani** [comment]
Scheduled for 2.10.2 through https://github.com/GeoNode/geonode/issues/5272
